Feature/init

ssysoenko opened the pull request.

shkryab committed:
Init

Add CMake helpers
Add factory

shkryab committed:
Add .clang-format

shkryab committed:
Add config

shkryab committed:
Add find boost

shkryab committed:
Add notice for boost

efimov-evgeniy committed:
Fix bug, add Config::makeFromFile, add test (#1)

* Fix bug, add Config::makeFromFile, add test

* Make name of test more meaningful

shkryab committed:
Remove "_" postfix
Add getChildren by name to Config
Add fPIC

efimov-evgeniy commented on include/cvs/common/configbase.hpp:
```diff
@@ -0,0 +1,281 @@
+      cvs::common::ConfigStaticValue<type, name##_name, config_value_kind, default_type __VA_OPT__(, __VA_ARGS__)>; \
+                                                                                                                    \
+ public:                                                                                                            \
+  Config_static_type_##name ::ResultType name;                                                                      \
```
А куда делись постфиксы?

ssysoenko replied:
Канули в небытие. Вообще думал поднять этот вопрос.
Зачем нам постфиксы?

ssysoenko replied:
И конкретно в этом случае задаешь имя ты без постфикса, а получаешь с постфиксом. Это понимается быстро, но все же впечатление портит.

efimov-evgeniy replied:
Я за префиксы, но по текущему стилю у нас постфиксы.

ssysoenko replied:
Чота я в гугл глянул и не нашел там ни префиксов ни постфиксов

efimov-evgeniy replied:
К тому же, ты убрал только в одном месте, а для объектов оставил.

ssysoenko replied:
Ну это я недосмотрел, да

efimov-evgeniy replied:
![image](https://user-images.githubusercontent.com/74912981/111638408-58c11a00-880b-11eb-8346-d63b134cfb1c.png)

ssysoenko replied:
Если бы я это увидел раньше - то оно бы было в гулодоке :)
Таки я бы поднял вопрос о необходимости этой лабуды. Не очень понимаю на кой ляд нужна кастрированная венгерская нотация.

shkryab committed:
Remove postfix for object_name

Aldrog commented on CMakeLists.txt:
```diff
@@ -0,0 +1,60 @@
+cmake_minimum_required(VERSION 3.16)
+
+include(GNUInstallDirs)
```
Генерирует предупреждение:
```
  Unable to determine default CMAKE_INSTALL_LIBDIR directory because no
  target architecture is known.  Please enable at least one language before
  including GNUInstallDirs.
```
Надо сначала объявить `project` и добавить к нему `LANGUAGES CXX`

ssysoenko replied:
А `LANGUAGES CXX` обязательно? Вроде и без него обходится вполне норм.

Aldrog replied:
Он по умолчанию `C CXX`, так что в целом опционально. Хотя на мой взгляд выключить фичи, связанные с поддержкой C, и добавить большей явности лишним не будет.

Aldrog commented on cmake/CVSBuildNumber.cmake:
```diff
@@ -0,0 +1,7 @@
+    else()
+        set(BUILD_NUMBER ".${BUILD_NUMBER}")
+    endif()
+endmacro()
```
Странный скрипт, не очень понятно, как его предполагается использовать.

ssysoenko replied:
Валялся у меня в общей куче и перешел сюда. Предполагается, что номер билдся добавляется только если задан снаружи например 
`cmake -DBUILD_NUMBER=5`
Ниже в проекте делается
`project(A VERSION 1.2.3${BUILD_NUMBER}`
Если параметр отсутствует - номер сборки просто не ставится.

Можно напихать туда хелперов для поддержки номера сборки который мы обсуждали. Но там есть вопросы про автоматизацию.

Aldrog replied:
Ага, теперь понятно. Может, тогда отдельные переменные для входа и выхода использовать? Например, результат записывать в `VERSION_BUILD_NUMBER`.

ssysoenko replied:
Да я вот думаю а не сделать ли отдельный макрос типа cvsproject. Возможно так еще правильнее будет.

Aldrog replied:
Да, кажется разумным. И `HOMEPAGE_URL` можно автоматом проставлять.

ssysoenko replied:
Обертка над `project` выдает предупреждение:
`CMake Warning (dev) in CMakeLists.txt:
  No project() command is present.  The top-level CMakeLists.txt file must
  contain a literal, direct call to the project() command.  Add a line of
  code such as
    project(ProjectName)
  near the top of the file, but after cmake_minimum_required().
  CMake is pretending there is a "project(Project)" command on the first
  line.
This warning is for project developers.  Use -Wno-dev to suppress it.`

Согласны мы с таким мириться? Причем он так ругается на макрос. Похоже cmake не умеет во вложенность и т.п.

ssysoenko replied:
Еще можно переопределить `project`. Оно даже работает. Варнинга нет но нативная функция переопределена.

Aldrog commented on include/cvs/common/configbase.hpp:
```diff
@@ -0,0 +1,281 @@
+
+#define SEARCH_IN_GLOBAL true
+
+#define VALUE(name, type, ...) \
+  HELPER_VALUE_BASE(name, type, cvs::common::ConfigValueKind::BASE, , void, __VA_OPT__(__VA_ARGS__))
+
+#define VALUE_OPTIONAL(name, type, ...) \
+  HELPER_VALUE_BASE(name, type, cvs::common::ConfigValueKind::OPTIONAL, , void, __VA_OPT__(__VA_ARGS__))
+
+#define VALUE_DEFAULT(name, type, default_value, ...)                                    \
+  HELPER_VALUE_BASE(                                                                     \
+      name, type, cvs::common::ConfigValueKind::WITH_DEFAULT_VALUE,                      \
+      struct Default_value_##name_type { static constexpr auto value = default_value; }; \
+      , Default_value_##name_type, __VA_OPT__(__VA_ARGS__))
```
Может, всё-таки добавим к этим макросам префикс `CONFIG_`? Кажется, `VALUE` слишком легко может с чем-то ещё пересечься.

efimov-evgeniy replied:
Имхо тогда лучше какую-то аббревиатуру добавить, типа "CVSC_". Так будет короче, и меньше вероятность, что пересечётся.

Aldrog commented on include/cvs/common/configbase.hpp:
```diff
@@ -0,0 +1,281 @@
+                                        &Self::object_name>>>
+
+#define OBJECT_OPTIONAL(object_name, ...) HELPER_OBJECT_BASE(object_name, true, __VA_ARGS__)
+#define OBJECT(object_name, ...)          HELPER_OBJECT_BASE(object_name, false, __VA_ARGS__)
```
И здесь тоже лучше бы префикс добавить.

shkryab committed:
Add prefix to config macroses

shkryab committed:
Change prefix

shkryab committed:
Move cmake helpers to another cmake package

shkryab committed:
Overrload project cmake function
Add cvs_add_to_aliast helper
Add cvs_git_current_hash helper
Remove CVSBuildNumber.cmake

Aldrog commented on cmake/CVSCommon.cmake:
```diff
@@ -0,0 +1,78 @@
+  endif()
+endmacro()
+
+macro(project)
```
Я так понимаю, тут есть риск всё сломать, включив CVSCommon дважды. Если вдруг мы всё-таки решим собирать зависимости вместе с проектом, это станет большой проблемой.

shkryab committed:
Fix recursion for multiple includes CVSCommon
